### PR TITLE
Fixed memory leaks/valgrind issues when having a double shutdown, issue #565

### PR DESCRIPTION
**Pull request by jwillemsen**

### CODE REVIEW

**Comment by jwillemsen** [comment]
Fixes #565

**Comment by jwillemsen** [comment]
Singleton close fixes the reported leaks from for example the `tests/DCPS/KeyTest> ./run_test.pl  isbounded` test run

**Comment by jwillemsen** [comment]
Nothing to add to this PR

**Comment by jwillemsen** [comment]
retest this please

**Comment by mitza-oci** [comment]
The Jenkins server is not current available.  We are working on bringing things back online from a power outage last week.

**Comment by mitza-oci** [comment]
retest this please

**Comment by jwillemsen** [comment]
Ran tests/transport/error_handling locally on this branch and no runtime issues there

**Comment by mitza-oci** [comment]
The test segfaults consistently on the Jenkins server:

```#0  0x0000000000ff0900 in ?? ()
#1  0x00007f63bfa8fdde in ace_cleanup_destroyer (object=0xff2000, param=0x0)
    at /var/lib/jenkins/jobs/TAO/workspace/ACE/ace/Cleanup.cpp:32
#2  0x00007f63bfa9018f in ACE_OS_Exit_Info::call_hooks (this=0xfed268)
    at /var/lib/jenkins/jobs/TAO/workspace/ACE/ace/Cleanup.cpp:170
#3  0x00007f63bfaedd86 in ACE_Object_Manager::fini (this=0xfed250)
    at /var/lib/jenkins/jobs/TAO/workspace/ACE/ace/Object_Manager.cpp:727
#4  0x00007f63bfaeee5d in ACE_OS_Object_Manager::fini (this=0xfed2f0)
    at /var/lib/jenkins/jobs/TAO/workspace/ACE/ace/Object_Manager_Base.cpp:320
#5  0x00007f63bfaee9ed in ACE_OS_Object_Manager::~ACE_OS_Object_Manager (
    this=0xfed2f0, __in_chrg=<optimized out>)
    at /var/lib/jenkins/jobs/TAO/workspace/ACE/ace/Object_Manager_Base.cpp:121
#6  0x00007f63bfaeea6e in ACE_OS_Object_Manager::~ACE_OS_Object_Manager (
    this=0xfed2f0, __in_chrg=<optimized out>)
    at /var/lib/jenkins/jobs/TAO/workspace/ACE/ace/Object_Manager_Base.cpp:122
#7  0x00007f63bfaef14d in ACE_OS_Object_Manager_Manager::~ACE_OS_Object_Manager_Manager (this=0x7f63bfdbf660 <ACE_OS_Object_Manager_Manager_instance>,
    __in_chrg=<optimized out>)
    at /var/lib/jenkins/jobs/TAO/workspace/ACE/ace/Object_Manager_Base.cpp:533
#8  0x00007f63bf10e53a in __cxa_finalize (d=0x7f63bfdbdcc0)
    at cxa_finalize.c:56
#9  0x00007f63bfa671d3 in __do_global_dtors_aux ()
   from /var/lib/jenkins/jobs/TAO/workspace/ACE/lib/libACE.so.6.4.3
#10 0x00007fffa6761e70 in ?? ()
#11 0x00007f63c0db170a in _dl_fini () at dl-fini.c:252
```

**Comment by jwillemsen** [comment]
Strange, can't reproduce it locally but probably the global cleanup wants to destroy the singleton again but we already closed it explicitly. The service participant now explicitly closes the singleton so it could be an ACE_Unmanaged_Singleton, testing this locally.

**Comment by jwillemsen** [comment]
@mitza-oci All tests are now running ok

File: dds/DCPS/Service_Participant.cpp [hunk 1]
```diff
@@ -253,9 +253,15 @@ Service_Participant::reactor_owner() const
...
 {
+  // When we are already shutdown just let the shutdown be a noop
+  if (shut_down_) {
+    return;
+  }
+
   shut_down_ = true;
   try {
     TransportRegistry::instance()->release();
+    TransportRegistry::instance()->close();
```
Comment (mitza-oci) [inline dds/DCPS/Service_Participant.cpp]: close() is declared static so invoke it as `TransportRegistry::close()`.  Also does close() need to be a distinct operation from release()?

Comment (jwillemsen) [inline dds/DCPS/Service_Participant.cpp]: Yes, should be `TransportRegistry::close()`. The release will let the transport registry release all its data, the close is for the singleton, should be two distinct operations

**Pull request merged**